src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PENamedTypeSymbol.cs

```diff
@@ -1804,9 +1892,185 @@ public override TypeKind TypeKind
                 }
 
                 return result;
+
+                bool mayBeExtension()
+                {
+                    return ReferenceEquals(_lazyDeclaredBaseType, ErrorTypeSymbol.UnknownResultType) ||
```

Comment:
>_lazyDeclaredBaseType [](http://example.com/codeflow?start=43&length=21)

It feels suspicious that we assume that ```_lazyDeclaredBaseType``` is initialized to something other than a sentinel value once we get here.

Reply:
In general, I think  we should use other (more robust) means to detect if the type is an extension. I understand that there is a circularity between ```TypeKind``` property and ```GetDeclaredBaseType``` method, but we should try to come up with an alternative way to break it.
For example, both ```TypeKind``` and  ```GetDeclaredBaseType``` could call a helper that initializes ```_lazyKind``` and returns it plus base produced by ```MakeBaseTypeFromMetadata``` at the same time.

Reply:
I'm not sure I see the problem, but I'm okay to compute both together.

Reply:
> I'm not sure I see the problem, but I'm okay to compute both together.

I couldn't see a way to get in a situation where _lazyDeclaredBaseType is not ErrorTypeSymbol.UnknownResultType here other than a race. In my opinion it is not worth optimizing (avoiding ```MakeBaseTypeFromMetadata()``` call) in a race situation. The logic could be much simpler. I guess, the suggested refactoring isn't simple as well, and we could simply remove the conditional logic around ```MakeBaseTypeFromMetadata``` call instead. However, with the suggested refactoring I was trying to avoid calling ```MakeBaseTypeFromMetadata``` again in ```GetDeclaredBaseType``` right after we check the type kind.